pkg/controller/factory/base_controller.go

```diff
@@ -0,0 +1,114 @@
+package factory
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+// baseController represents generic Kubernetes controller boiler-plate
+type baseController struct {
+	cachesToSync    []cache.InformerSynced
+	sync            func(ctx context.Context, controllerContext SyncContext) error
+	resyncEvery     time.Duration
+	ctx             SyncContext
+	shutdownContext context.Context
+}
+
+var _ Controller = &baseController{}
+
+func (c *baseController) Run(ctx context.Context, workers int) {
+	shutdownContext, shutdownComplete := context.WithCancel(context.Background())
+	if c.shutdownContext != nil {
+		panic(fmt.Sprintf("controller %q is already running", c.ctx.ControllerName()))
+	}
+	c.shutdownContext = shutdownContext
+
+	defer shutdownComplete()
+	defer utilruntime.HandleCrash()
+	defer c.ctx.Queue().ShutDown()
+	defer klog.Infof("Shutting down %s ...", c.ctx.ControllerName())
+
+	klog.Infof("Starting %s ...", c.ctx.ControllerName())
+	if !cache.WaitForCacheSync(ctx.Done(), c.cachesToSync...) {
+		return
```

Comment:
should we have an error as result `Run`?

Reply:
we should panic in that case (we recover)